src/bgp/fsm/BGPStateOpenSent.cpp

```diff
@@ -227,44 +228,51 @@ bool BGPStateOpenSent ::onEvent(BGPEvent event) {
             // applied
             // when a valid BGP OPEN message is received (Event 19 or Event 20).
 
-            // FIXME check how to call collision detection
-            // stateMachine->connection->bgpApplication->collisionDetection(stateMachine->connection);
+            // DONE check how to call collision detection
 
+            {
+                std::unique_ptr<BGPOpenLayer> openLayer;
+                dynamic_pointer_move(openLayer, event.layers);
 
-            // resets the DelayOpenTimer to zero,
-            stateMachine->resetDelayOpenTimer();
+                BGPOpenLayer::BGPOpenHeader *openHeader =
+                    openLayer->getOpenHeaderOrNull();
 
-            // sets the BGP ConnectRetryTimer to zero,
-            stateMachine->resetConnectRetryTimer();
+                pcpp::IPv4Address bgpIdentifier =
+                    pcpp::IPv4Address(be32toh(openHeader->BGPIdentifier_be));
 
-            // sends a KEEPALIVE message, and
-            // TODO test KEEPALIVE message
-            {
-                std::unique_ptr<BGPLayer> bgpKeepaliveLayer =
-                    std::make_unique<BGPKeepaliveLayer>();
-                bgpKeepaliveLayer->computeCalculateFields();
+                stateMachine->connection->bgpApplication->collisionDetection(
+                    stateMachine->connection, bgpIdentifier);
 
-                std::unique_ptr<std::stack<std::unique_ptr<pcpp::Layer>>>
-                    layers =
-                        make_unique<std::stack<std::unique_ptr<pcpp::Layer>>>();
-                layers->push(std::move(bgpKeepaliveLayer));
 
-                stateMachine->connection->sendData(std::move(layers));
-                L_INFO(stateMachine->connection->owner->ID + " " +
-                           stateMachine->name,
-                       "Sending KEEPALIVE message");
-            }
+                // resets the DelayOpenTimer to zero,
+                stateMachine->resetDelayOpenTimer();
 
-            // sets a KeepaliveTimer (via the text below)
+                // sets the BGP ConnectRetryTimer to zero,
+                stateMachine->resetConnectRetryTimer();
 
-            // sets the HoldTimer according to the negotiated value (see
-            // Section 4.2),
-            {
-                std::unique_ptr<BGPOpenLayer> openLayer;
-                dynamic_pointer_move(openLayer, event.layers);
+                // sends a KEEPALIVE message, and
+                // TODO test KEEPALIVE message
```

Comment:
What if the collision detection marks this connection as the one to be closed? After the decision, we are sending a keepalive regardless of the decision. Is it the expected behavior?